Alerting: Allow sending notification tags to Opsgenie as extra properties

DewaldV opened the pull request:
**What this PR does / why we need it**: 

Allow users to select where to send notification tags when alerting via OpsGenie. 

Users will be able to select where to send the notification tags and then see those tags as key/values under extra properties or tags in an alert on the Opsgenie UI. This metadata can then be further used on the Opsgenie platform. 

Default will be the current behaviour (called `Tags` in the dropdown) with an option to send to both if users want to keep the current behaviour while moving new config to using extra properties.

  * Configurable delivery to either tags, extra properties or both

  * Default to current behaviour (tags only)

  * Support both so users can transition from tags to details

**Which issue(s) this PR fixes**:
Fixes #30331

**Special notes for your reviewer**:

I have added docs for this feature and also one that previously wasn't documented (`og_priority`) but would appreciate a copy-edit of the wording in case.

Ran all the tests locally and everything looked good, also have some screenshots of the final result on the Opsgenie side that I used for issue #30331.

I did read through the style guide but opted to keep this change using `SimpleJson` and `GoConvey` as there was already usage of it in the notifier and I didn't think this PR was suitable for a full refactor to either. I might be keen to submit a PR for that kind of refactor in the future though if you're looking for those kinds of submissions too.

DewaldV committed:
Alerting: Opsgenie send tags as extra properties

Allow users to select where to send notification tags when alerting via
OpsGenie. Supports sending tags as key/value details, concatenated
strings in tags or both.

Users will be able to see their tags as key/values under extra
properties in an alert on the Opsgenie UI. These key/values can
then be used in the platform for routing, templating etc.

  * Configurable delivery to either tags, extra properties or both

  * Default to current behaviour (tags only)

  * Support both so users can transition from tags to details

Add docs and clean up references

DewaldV commented:
CI seems to have failed on a single end to end test.

```
1) adding a multi value query variable
...
AssertionError: Timed out retrying: Expected to find element: `1`, but never found it. Queried from element: <a.css-10l6kcd>
```

I can see this has failed before on [another unrelated build](https://drone.grafana.net/grafana/grafana/10001/1/20) so I suspect this might be a flaky test. Would appreciate if anyone with CI access could give the build a restart. :pray: 

If there's a command to have a bot tackle the restart I'm happy to learn too. :+1:

oddlittlebird commented:
Thanks for the contribution! We are currently in the Grafana 7.4 beta period, so I don't know if anyone will have time to review this until after that.

If you see a notification from Stalebot, just comment on the PR to make it go away.

DewaldV commented:
Thanks @oddlittlebird, appreciate I came in at an awkward time for the release. :sweat_smile: 

I'll keep an eye out and pop a comment here if needed.

Good luck with the 7.4 Beta release! :+1:

DewaldV commented:
Hi @oddlittlebird. Just checking in but now that Grafana 7.4 has been released do you think we could put this one back in the queue for a review?

oddlittlebird commented:
That's up to @kylebrandt and @ying-jeanne.

DewaldV commented:
Rebased onto latest master to ensure we're up to date with any incoming changes. 

Still awaiting a review when anyone has a moment. :smile:

DewaldV commented:
Whoops, rebase seems to have introduced some issues. I'll pick these up and get the build back to green.

DewaldV committed:
Alerting: Add additional arg for Opsgenie tests

The NewEvalContext function now requires a 'PluginRequestValidator' argument.
As our test does not use the validator we can specify 'nil' to satisfy the
function and allow our test to pass as expected.

DewaldV commented:
Ok, pushed a fix for the issue after the first rebase and then had to rebase again since the first rebase seemed to have missing files for CI. 

2nd run is looking much better now. :+1:

marefr reviewed (changes requested):
Looks good. But this documentation page needs an update as well in regards to opsgenie: 
- https://grafana.com/docs/grafana/latest/administration/provisioning/#supported-settings
- https://github.com/grafana/grafana/blob/master/docs/sources/administration/provisioning.md#alert-notification-opsgenie

achatterjee-grafana commented on docs/sources/alerting/notifications.md:
```diff
@@ -111,6 +111,19 @@ Token | If provided, Grafana will upload the generated image via Slack's file.up
+Setting | Description
+--------|------------
+Alert API URL | The API URL for your Opsgenie instance. This will normally be either `https://api.opsgenie.com` or, for EU customers, `https://api.eu.opsgenie.com`.
+API Key | The API Key as provided by Opsgenie for your configured Grafana Integration.
```
```suggestion
API Key | The API Key as provided by Opsgenie for your configured Grafana integration.
```

achatterjee-grafana commented on docs/sources/alerting/notifications.md:
```diff
@@ -111,6 +111,19 @@ Token | If provided, Grafana will upload the generated image via Slack's file.up
+--------|------------
+Alert API URL | The API URL for your Opsgenie instance. This will normally be either `https://api.opsgenie.com` or, for EU customers, `https://api.eu.opsgenie.com`.
+API Key | The API Key as provided by Opsgenie for your configured Grafana Integration.
+Override priority | Allows the alert priority to be set using the `og_priority` tag. The `og_priority` tag must have one of the following values: `P1`, `P2`, `P3`, `P4` or `P5`. Default is `False`.
```
```suggestion
Override priority | Configures the alert priority using the `og_priority` tag. The `og_priority` tag must have one of the following values: `P1`, `P2`, `P3`, `P4,` or `P5`. Default is `False`.
```

achatterjee-grafana commented on docs/sources/alerting/notifications.md:
```diff
@@ -111,6 +111,19 @@ Token | If provided, Grafana will upload the generated image via Slack's file.up
+Alert API URL | The API URL for your Opsgenie instance. This will normally be either `https://api.opsgenie.com` or, for EU customers, `https://api.eu.opsgenie.com`.
+API Key | The API Key as provided by Opsgenie for your configured Grafana Integration.
+Override priority | Allows the alert priority to be set using the `og_priority` tag. The `og_priority` tag must have one of the following values: `P1`, `P2`, `P3`, `P4` or `P5`. Default is `False`.
+Send notification tags as | Allows you to specify how you would like [Notification Tags]({{< relref "create-alerts.md/#notifications" >}}) delivered to Opsgenie. They can be delivered as `Tags`, `Extra Properties` or both. Default is `Tags`. (1)
```
```suggestion
Send notification tags as | Specify how you would like [Notification Tags]({{< relref "create-alerts.md/#notifications" >}}) delivered to Opsgenie. They can be delivered as `Tags`, `Extra Properties` or both. Default is `Tags`. (1)
```

achatterjee-grafana commented on docs/sources/alerting/notifications.md:
```diff
@@ -111,6 +111,19 @@ Token | If provided, Grafana will upload the generated image via Slack's file.up
+Override priority | Allows the alert priority to be set using the `og_priority` tag. The `og_priority` tag must have one of the following values: `P1`, `P2`, `P3`, `P4` or `P5`. Default is `False`.
+Send notification tags as | Allows you to specify how you would like [Notification Tags]({{< relref "create-alerts.md/#notifications" >}}) delivered to Opsgenie. They can be delivered as `Tags`, `Extra Properties` or both. Default is `Tags`. (1)
+
+> **Note:** When notification tags are sent as `Tags` they will be concatenated into string with a `key:value` format. If you prefer to receive the notifications tags as key/values under Extra Properties in Opsgenie then change the `Send notification tags as` to either `Extra Properties` or `Tags & Extra Properties`.
```
```suggestion
> **Note:** When notification tags are sent as `Tags` they are concatenated into a string with a `key:value` format. If you prefer to receive the notifications tags as key/values under Extra Properties in Opsgenie then change the `Send notification tags as` to either `Extra Properties` or `Tags & Extra Properties`.
```

achatterjee-grafana reviewed (changes requested):
Added some copy-edit suggestions.

DewaldV commented:
Thanks for the reviews both.

@marefr I'll get those additional pages updated as well.

@achatterjee-grafana Appreciate the changes, I'll batch them up and accept. :+1:

DewaldV committed:
Alerting: Opsgenie doc fixes

Accept suggested changes for Opsgenie docs

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

marefr commented:
@DewaldV sorry if you're already on your way fixing this one, but still needs an update:
https://github.com/grafana/grafana/blob/master/docs/sources/administration/provisioning.md#alert-notification-opsgenie

DewaldV committed:
Alerting: Opsgenie provisioning settings docs

Add the new setting to the provisioning docs

DewaldV commented:
@marefr Apologies I pushed them as separate commits as I wanted to ensure I keep @achatterjee-grafana's suggestions as its own commit.

I also wanted to rebase onto master to ensure there aren't any merge issues so thought I might as well do a commit for your requested change specifically.

DewaldV committed:
Alerting: Opsgenie doc typo correction

Move the comma (,) out of the preformatting tags for the setting value

DewaldV committed:
Alerting: Opsgenie refactor send switches

Refactor the send switches to be methods on the OpsgenieNotiefier
itself. This also cleans up the method names so that the code reads
a bit more logically as:

if we should send details: send details
if we should send tags: send tags

This avoids the calling code needing to care about passing the state
and allows an engineer working in the `createAlert` function to focus
on the results of applying the logic instead.

marefr commented on docs/sources/alerting/notifications.md:
```diff
@@ -111,6 +111,19 @@ Token | If provided, Grafana will upload the generated image via Slack's file.up
+Alert API URL | The API URL for your Opsgenie instance. This will normally be either `https://api.opsgenie.com` or, for EU customers, `https://api.eu.opsgenie.com`.
+API Key | The API Key as provided by Opsgenie for your configured Grafana integration.
+Override priority | Configures the alert priority using the `og_priority` tag. The `og_priority` tag must have one of the following values: `P1`, `P2`, `P3`, `P4`, or `P5`. Default is `False`.
+Send notification tags as | Specify how you would like [Notification Tags]({{< relref "create-alerts.md/#notifications" >}}) delivered to Opsgenie. They can be delivered as `Tags`, `Extra Properties` or both. Default is `Tags`. (1)
```
What is `(1)` referring to at the end here? This is the only thing left as I see to get this merged.

DewaldV replied:
@marefr It's the first note below. I took inspiration from the PagerDuty docs that used the same mechanism to refer to notes below. Maybe it would be a bit clearer if I added the `(1)` to the Note itself to link it back and made a link. Give me a few minutes to get that done.

marefr replied:
Hmm okay. Not sure. Maybe keeping the note and refer to "Default is `Tags`. See note below for more information.". Lets ask @achatterjee-grafana for input on what make most sense.

achatterjee-grafana replied:
Agree with @marefr. This is not consistent with our style guide and use. It would be confusing to the reader, as evident from @marefr's reaction. Let's change it as per the recommendation!

DewaldV replied:
Can do I'll push another fix.

DewaldV committed:
Alerting: Opsgenie docs rename note


Rename the note heading to match the number to more clearly link them.

DewaldV committed:
Alerting: Opsgenie use standard reference to note

Refer to the note below as per recommendation and standards.

achatterjee-grafana reviewed (approved):
LGTM

marefr reviewed (approved):
LGTM

marefr commented:
Thank you for contributing to Grafana!

DewaldV commented:
Thanks for the reviews, comment and approbals @marefr and @achatterjee-grafana :tada:
